stl/inc/xstring

```diff
@@ -2386,7 +2391,7 @@ public:
     }
 
     union _Bxty { // storage for small buffer or pointer to larger one
-        _CONSTEXPR20 _Bxty() noexcept : _Ptr() {} // user-provided, for fancy pointers
+        _CONSTEXPR20 _Bxty() noexcept : _Buf() {} // user-provided, for fancy pointers

```

Comment:
Yes this is indeed necessary. Otherwise the union would have no active member. Note that when going small we simply default construct `_Bxty` which initializes the SSO block

Reply:
Gotcha -- Is there a `constexpr` reason that this is necessary now, but it was okay for `_Bxty` to not have an initial active member before we reached in and messed with things for WG21-P0980?  Is this one of those "technically UB but it was fine but now we're `constexpr` so UB is (channeling @CaseyCarter here) _verboten_?"

Reply:
Leaving the union with no active member should be fine per C++20 `constexpr` rules thanks to WG21-P1331. (I have no idea about the implementation status in our various compilers of P1331 or the "assignment can activate a union member" change whose proposal I don't recall.)

Reply:
Ah now I remember it, the issue is constructing the array in place. With that constructor I can simply write
```cpp
_Construct_in_place(_Mypair._Myval2._Bx);
```

And the SSO buffer is fully set up and life times are handled. Otherwise I need to properly construct a C array via `_Construct_in_place` and that gives me nightmares

Reply:
We should be able to force `_Buf` into existence by assigning to one of its elements (`_Buf[0] = _Elem{}`, probably?) via the magic of https://eel.is/c++draft/class.union.general#6.

Reply:
I think that rule is bogus. You cannot assign to something whose livetime has not startet yet

Reply:
> I think that rule is bogus. You cannot assign to something whose livetime has not startet yet

Perhaps that's why "the beginning of its lifetime is sequenced after the value computation of the left and right operands and before the assignment"?

Reply:
As guessed his is not yet supported by compilers. It actually seem to be a proper warning:
```
C:\STL\out\build\x64\out\inc\xstring(52,9): note: assignment to member '_Buf' of union with active member '_Ptr' is not allowed in a constant expression
        __builtin_memcpy(_First1, _First2, _Count * sizeof(_Elem));
```

So I guess I am writing a bug report

Reply:
FYI, I tried to reproduce this but was not immediately able to. Will try later with a more complex setting

Reply:
after talking about it more on discord, we came to the conclusion that this is actually an MSVC bug.
https://discordapp.com/channels/737189251069771789/737189251497721887/869117117683867659

https://discordapp.com/channels/737189251069771789/737189251497721887/869255296718102558